### ISSUE TITLE
Dont interpret coordinates relative to dataset bounding box

### ISSUE THREAD

**Issue by philippotto** [scalableminds/webknossos-cuber#234]
I just debugged some code of mine and finally found out that the high-level dataset api does not directly use the passed offsets (for reading and writing). Instead, the offset is interpreted relative to the global bounding box which is specified in the datasource-properties.json. I think this is problematic and not the behavior which is expected in most cases.

For the "Views" which are bounded to a specific bounding box, it seems reasonable to stick to the "relativeness", but when accessing the entire dataset, I'd expect to use the global coordinate system. Do you agree, @normanrz @rschwanhold ? This would also align with the "old wkw api access".

- [x] discuss
- [x] double-check whether existing code already relied on the old behavior
- [x] change implementation

**Comment by rschwanhold** [comment]
I wasn't aware, that the "old wkw api" did it differently. I agree with you that we should change this.

That raises the question for me: What is the purpose of the offset in the properties.json? Are there other use-cases than to restrict a view?

**Comment by philippotto** [comment]
> I agree with you that we should change this.

Great! Could you do this? :)

> That raises the question for me: What is the purpose of the offset in the properties.json? Are there other use-cases than to restrict a view?

In general, it's used to know "where the data is". Either in webKnossos when viewing, but also when doing automatic processing (e.g, segmenting the data) it's used to read the correct bounding box.

**Comment by rschwanhold** [comment]
@philippotto I would like to get your opinion on how the bounding box of views should be calculated. (Maybe I gave too much attention to the offset in the properties.json)
- Current behavior:
  - the `MagDataset` has a `View` which has its origin at the offset that is defined in the properties.json
  - `MagDataset.get_view` takes two optional parameters (`size` and `offset`) which are combined with the size and offset from the properties.json
    - the offset of the new view is the addition of the offset from the properties and the passed parameter
    - if the size is specified as a parameter, that size is used for the new view (otherwise the value from the properties is used)
  - A `View` itself also has a function `get_view` (to get a sub-view). The offset of the new view is relative to the view that created it

- Possible new behavior:
  - the `MagDataset` has a `View` which has its origin always at (0, 0, 0)
  - `get_view` takes two optional parameters (`size` and `offset`) 
    - if the parameters `size` and `offset` are not specified, the size and offset from the properties.json are used
    - what is the expected behavior if the parameters are specified?
      - should the offset be interpreted relative to (0, 0, 0) or to the offset from the properties.json?
        - in case we interpret the parameter as an absolute value (relative to (0, 0, 0)): should it be allowed to create a view with `is_bounded=True` (which is the default value) that is outside the bounding box in the properties.json?
  - I would like to leave the `View.get_view` as it is

**Comment by philippotto** [comment]
> Possible new behavior:
> 
> * the `MagDataset` has a `View` which has its origin always at (0, 0, 0)
> * `get_view` takes two optional parameters (`size` and `offset`)
>   * if the parameters `size` and `offset` are not specified, the size and offset from the properties.json are used
>   * what is the expected behavior if the parameters are specified?
>     * should the offset be interpreted relative to (0, 0, 0) or to the offset from the properties.json?
>       (I prefer to interpret the offset as an absolute value (relative to (0, 0, 0)))
>       * in case we interpret the parameter as an absolute value: should it be allowed to create a view with `is_bounded=True` (which is the default value) that is outside the bounding box in the properties.json?
> * I would like to leave the `View.get_view` as it is

In general, this approach sounds very good :+1: 


> - what is the expected behavior if the parameters are specified?
>   - should the offset be interpreted relative to (0, 0, 0) or to the offset from the properties.json? (I prefer to interpret the offset as an absolute value (relative to (0, 0, 0)))

Yes, I'd interpret it relative to (0, 0, 0) (so, it's basically absolute).

> - in case we interpret the parameter as an absolute value: should it be allowed to create a view with is_bounded=True (which is the default value) that is outside the bounding box in the properties.json?

`is_bounded` basically means that one cannot write outside of the existing bounding box, right? Then, yes, one could forbid creating a bounded view outside of the bounding box (by throwing an exception). But I think it's also fine to defer the exception until one really tries to read or write (this should already work then, if I'm not mistaken).

-----------

To summarize: All the offsets should be interpreted in absolute terms with the only exception of using a sub view (if that would be absolute, too, the functionality wouldn't make much sense), am I right?

**Comment by rschwanhold** [comment]
> To summarize: All the offsets should be interpreted in absolute terms with the only exception of using a sub view (if that would be absolute, too, the functionality wouldn't make much sense), am I right?

I agree.

We have now only talked about the `offset`-parameter for `get_view`.
A View has the method `write(self, data, offset=(0, 0, 0))`. Are we on the same page regarding the offset of that method? I interpret the `offset` of the `write`-method as a relative value to the offset of the view. 
This means, that calling `write` on a `MagDataset` will always be "reative" to (0, 0, 0) (as discussed above). 
(Maybe I should rename the parameter from the `write`-method to `relative_offset`?!)

**Comment by philippotto** [comment]
Yes, I agree with everything you said. `relative_offset` for a view sounds reasonable, but for MagDataset it would be confusing (since it will always be relative to 0, 0, 0 which is basically absolute). So, maybe it's best to use different parameter names for MagDataset and View. The idea seemed a bit unelegant to me at first, but I think it's a good way of ensuring that the user of the code is aware of the difference.
